### DISCUSSION TITLE
How can I encode an ABI for a struct with an array of structs inside of it?

### DISCUSSION THREAD

**Discussion by 0xPT** [wevm/viem#868] [Question]
I've got what I'm trying to do working in Ethers.js with their AbiCoder.

['address', 'tuple(bytes32, bytes32)[]', 'uint256', 'uint256'] is the abiParams that work in Ethers.

If I try to use the same types but formatted properly for the encodeAbiParameters I either get this error

`InvalidAbiEncodingType: Type "tuple(bytes32, bytes32)" is not a valid encoding type.`
Please provide a valid ABI type.

or this error when I change it to just be tuple[]
`TypeError: Cannot read properties of undefined (reading 'length') at encodeTuple`

**Comment by jxom** [comment]
Just remove the `tuple` keyword so it’s: `(bytes32, bytes32)`

**Reply by 0xPT** [reply]
InvalidAbiEncodingType: Type "(bytes32, bytes32)" is not a valid encoding type.

**Reply by 0xPT** [reply]
```
encodeAbiParameters(
    [
      { name: "address", type: "address" },
      { name: "structArray", type: "type[bytes32, bytes32]" },
      { name: "num1", type: "uint256" },
      { name: "num2", type: "uint256" },
    ],
  );
```

Would it make sense to just encode structArray first and change it here to be "bytes", like so?

```
const encodedStructArray = encodeAbiParameters(
    [
      {
        name: "structArray",
        type: [{ name: "val1", type: "bytes32" }, { name: "val2", type: "bytes32" }]
      }
    ],
    [[{val1: "bar", val2: "foo"}, {val1: "foo", val2: "bar"}]]
)
encodeAbiParameters(
    [
      { name: "address", type: "address" },
      { name: "structArray", type: "bytes" },
      { name: "num1", type: "uint256" },
      { name: "num2", type: "uint256" },
    ],
    [address, encodedStructArray, num1, num2]
  );
```

If so, the only problem I would still face is not understanding how to encode an array properly (gave it my best shot with "encodedStructArray", getting some type errors and runtime errors on that though as well)

**Reply by jxom** [reply]
Oh, yeah that type isn’t a valid ABI type. You’d need to pass a `tuple` type with its parameters as `components`. If you want to do the human readable approach, you can use `parseAbiParameters`: https://viem.sh/docs/abi/encodeAbiParameters.html#human-readable
